File: src/GiftExchange.Library/Models/DeliveryStatus.cs
namespace GiftExchange.Library.Models;

/// <summary>
/// How far a participant email is known to have got, as SES reported it.
///
/// The empty string is a real value here and the one most rows start at: it means nothing has been
/// heard yet, which is not the same as "not delivered". Everything that reads this has to keep the
/// two apart — an organizer told somebody has not received their invitation, when in truth no event
/// has arrived, will go and pester a person who is holding it.
/// </summary>
public static class DeliveryStatus
{
    /// <summary>Nothing has been heard about this participant. Not a status SES ever reports.</summary>
    public static string Unknown => string.Empty;

    /// <summary>SES accepted the message. It has left this application and nothing more is known.</summary>
    public static string Sent => "SENT";

    /// <summary>Temporarily undeliverable and still being retried. Neither a success nor a failure yet.</summary>
    public static string Delayed => "DELAYED";

    /// <summary>The receiving mail server accepted it. The furthest anything here can honestly claim.</summary>
    public static string Delivered => "DELIVERED";

    /// <summary>The recipient marked it as spam. Arrived, and was unwelcome.</summary>
    public static string Complained => "COMPLAINED";

    /// <summary>SES refused to send it, typically because the content tripped its virus scan.</summary>
    public static string Rejected => "REJECTED";

    /// <summary>SES could not render the message at all, so nothing was sent.</summary>
    public static string Failed => "FAILED";

    /// <summary>It came back. The one status an organizer can actually do something about.</summary>
    public static string Bounced => "BOUNCED";
}

public static class DeliveryStatuses
{
    public static readonly ImmutableList<string> All =
    [
        DeliveryStatus.Sent,
        DeliveryStatus.Delayed,
        DeliveryStatus.Delivered,
        DeliveryStatus.Complained,
        DeliveryStatus.Rejected,
        DeliveryStatus.Failed,
        DeliveryStatus.Bounced
    ];

    /// <summary>
    /// How far through the message's life each status sits, so that a row only ever moves forwards.
    /// </summary>
    /// <remarks>
    /// SES does not order event delivery, and SNS does not either. A Delivery published before a
    /// Send can be handed to us after it, and a row that took the last event to arrive rather than
    /// the furthest one reached would flap between DELIVERED and SENT for no reason anybody could
    /// see in the data.
    ///
    /// The bad outcomes rank above Delivered rather than beside it, deliberately. A complaint is
    /// something the recipient does after the message arrived, so it has to be able to overwrite
    /// the delivery it followed; a bounce outranks everything, because a message that came back is
    /// the fact worth keeping no matter what else was said about it.
    /// </remarks>
    public static int RankOf(string status) => status switch
    {
        var value when value == DeliveryStatus.Sent => 1,
        var value when value == DeliveryStatus.Delayed => 2,
        var value when value == DeliveryStatus.Delivered => 3,
        var value when value == DeliveryStatus.Complained => 4,
        var value when value == DeliveryStatus.Rejected => 5,
        var value when value == DeliveryStatus.Failed => 5,
        var value when value == DeliveryStatus.Bounced => 6,
        // Unknown, and anything SES invents after this was written. Ranking it below every real
        // status means a status we do not recognise can never displace one we do.
        _ => 0
    };
}

File: src/app/src/deliveryStatus.ts
/**
 * User-facing labels for what SES said happened to a participant's email.
 *
 * The wire values are the server's vocabulary and stay as they are; this is the only place the UI
 * decides what to call them, for the reason hatStatus.ts gives.
 *
 * The empty status is the one that matters most to get right. It means nothing has been heard yet
 * — which is the state for the first seconds after a send, and the permanent state for a message
 * whose events were lost — and it is emphatically not "this person did not receive it". An
 * organizer who reads it as the latter goes and pesters somebody who is holding their invitation,
 * so the label says what is true about our knowledge rather than about the participant.
 */
const DELIVERY_STATUS_LABELS: Record<string, string> = {
  '': 'No confirmation yet',
  SENT: 'Sent',
  DELAYED: 'Delayed, still trying',
  DELIVERED: 'Delivered',
  COMPLAINED: 'Marked as spam',
  REJECTED: 'Rejected',
  FAILED: 'Failed to send',
  BOUNCED: 'Bounced',
}

/**
 * What to call each of the things this application sends, when saying which one a status is about.
 *
 * An exchange sends more than one message to the same person and the status shown is the newest of
 * them, so an organizer reading "Delivered" beside somebody who says they never saw their
 * invitation is owed the answer to "delivered what". After an exchange closes that word is often
 * about the announcement, sent weeks after the invitation that went astray.
 *
 * INVITATION and COMPLETION borrow the words EditAddressModal already uses for the same two
 * emails. The two leave notices are named as one thing on purpose: they are the same event told to
 * different people, and the distinction between them is ours rather than the reader's.
 *
 * UNSPECIFIED is deliberately absent. It means a send reached SES without a type tag, which is a
 * mistake in this codebase rather than a fact about the participant, and naming it in the table
 * would ask an organizer to make sense of it.
 */
const MESSAGE_TYPE_LABELS: Record<string, string> = {
  INVITATION: 'Invitation',
  COMPLETION: 'Announcement',
  PARTICIPANT_LEFT: 'Someone-left notice',
  ORGANIZER_LEFT_NOTE: 'Someone-left notice',
}

/** The empty string for anything unnamed above, which is the signal to render nothing. */
export function formatDeliveryMessageType(messageType: string): string {
  return MESSAGE_TYPE_LABELS[messageType] ?? ''
}

/**
 * Which statuses the organizer can actually do something about.
 *
 * Only these show the detail underneath. For the rest the detail is either empty or says nothing
 * an organizer would act on, and a row of explanations nobody needs would bury the one that
 * matters.
 */
const ACTIONABLE_STATUSES = ['BOUNCED', 'REJECTED', 'FAILED']

/** Grouped for colour: good, waiting, or wrong. */
export type DeliveryTone = 'good' | 'neutral' | 'bad'

export function deliveryTone(status: string): DeliveryTone {
  if (status === 'DELIVERED') return 'good'
  if (ACTIONABLE_STATUSES.includes(status) || status === 'COMPLAINED') return 'bad'
  return 'neutral'
}

export function formatDeliveryStatus(status: string): string {
  return (
    DELIVERY_STATUS_LABELS[status] ??
    status
      .split('_')
      .map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
      .join(' ')
  )
}

/**
 * Whether the reason underneath is worth showing.
 *
 * The detail itself is written by whatever mail server rejected the message — it passes through no
 * moderation on the way here, ours or anybody's. Rendering it as React text rather than as markup
 * is what keeps that safe.
 */
export function showsDeliveryDetail(status: string, detail: string): boolean {
  return detail.length > 0 && ACTIONABLE_STATUSES.includes(status)
}
